README.md
# ZenWeb Body module

[ZenWeb](https://www.npmjs.com/package/zenweb)

支持 JSON、Form 等内容格式的基础解析模块

## 演示
```ts
import { Context, mapping, Body, BodyHelper } from 'zenweb';

export class Controller {
  @mapping({ path: '/', method: 'POST' })
  post(body: Body) {
    console.log(body.type); // POST body 内容类型
    console.log(body.data); // POST Body 内容解析完成后的数据
  }

  @mapping({ path: '/', method: 'POST' })
  post(body: BodyHelper) {
    console.log(body.get({ age: '!int' })); // 类型转换&校验
  }
}
```

## 内置的对象内容解析器

- json
- x-www-form-urlencoded

## 配置项

| 项 | 类型 | 默认值 | 说明 |
|----|-----|-------|-----|
| encoding | string | 'utf-8' | 在未能匹配到 header 信息时，默认text内容编码格式 |
| limit | number | 1MB | 内容大小限制 |
| inflate | boolean | true | 是否支持压缩内容解压 |
| textTypes | string[] | ['text/*', 'json', '+json', 'xml', '+xml', 'urlencoded'] | 支持解析为文本的类型 |
| parses | BodyParserClass[] | [JSONParser, URLEncodedParser] | 对象解析器 |

## 其他扩展格式解析

[xml](https://www.npmjs.com/package/@zenweb/xml-body)


CHANGELOG.md
# Changelog

## [3.6.1] - 2023-3-24
- fix: 不使用 node16 的包导入格式

## [3.6.0] - 2023-3-19
- update: Body 不识别的类型不再解析为 raw。对于文件上传和Body共同使用的场景下，一般来说是不需要再读取流的，浪费内存，如果真的需要直接使用 RawBody 即可

## [3.5.1] - 2023-3-19
- fix: TextBody 判断顺序优化，先判断是否为支持类型，不支持的直接忽略

## [3.5.0] - 2023-3-17
- 内容解析器使用依赖注入管理

## [3.4.0] - 2023-3-17
- 可扩展的内容解析器

## [3.3.1] - 2023-3-16
- fix: 空内容体支持
- update: JSON 解析错误捕获

## [3.3.0] - 2023-3-16
- 重写内容解析逻辑
- 新增: TextBody 和 RawBody 注入类型
- 修改: 配置项简化，不再使用 ctx.fail 输出错误，而是使用标准 httpError

## [3.2.2] - 2023-3-15
- update: application/xml to text

## [3.2.1] - 2023-3-14
- 修改: ObjectBody, Body.data 对象项目类型为 unknown

## [3.2.0] - 2023-3-13
- 新增: ObjectBody 对象
- 改进: 错误消息使用 messagecode 管理

## 3.1.1
- null 检查

## 3.1.0
- 新增 BodyHelper 数据类型转换&校验

## 3.0.0
使用依赖注入重构，去除 xml 和 文件上传表单支持，取消的这两个作为独立模块分离。


test/t1.ts
import { Core } from '@zenweb/core';
import inject from '@zenweb/inject';
import result from '@zenweb/result';
import body, { Body } from '../src';

const app = new Core();
app.setup(result());
app.setup(inject());
app.setup(body());
app.setup(function test(setup) {
  setup.middleware(async ctx => {
    const body = await ctx.injector.getInstance(Body);
    ctx.body = body;
  });
})
app.start();


src/read.ts
import * as httpError from 'http-errors';
import * as zlib from 'zlib';
import { Readable } from 'stream';
import { Context } from '@zenweb/core';

/**
 * 内容读取流
 * @param ctx 
 * @param inflate 是否启用解压缩支持
 */
export function streamReader(ctx: Context, limit?: number, inflate = true) {
  // 长度检查
  const length = ctx.request.length;
  if (limit && length && length > limit) {
    throw httpError(413, 'request entity too large', {
      expected: length,
      length: length,
      limit: limit,
      type: 'entity.too.large',
    })
  }

  // 内容格式检查
  const encoding = (ctx.get('content-encoding') || 'identity').toLowerCase();
  if (inflate === false && encoding !== 'identity') {
    throw httpError(415, 'content encoding unsupported', {
      encoding: encoding,
      type: 'encoding.unsupported',
    });
  }

  let stream: Readable & { length?: number };
  const req = ctx.req;

  if (encoding === 'identity') {
    stream = req;
    stream.length = length;
  } else if (encoding === 'deflate') {
    let s = zlib.createInflate();
    req.pipe(s);
    stream = s;
  } else if (encoding === 'gzip') {
    let s = zlib.createGunzip();
    req.pipe(s);
    stream = s;
  } else {
    throw httpError(415, 'unsupported content encoding "' + encoding + '"', {
      encoding: encoding,
      type: 'encoding.unsupported',
    });
  }

  if (!stream.readable) {
    throw httpError(500, 'stream is not readable', {
      type: 'stream.not.readable',
    });
  }

  return new Promise<Buffer>((resolve, reject) => {
    let complete = false;
    let received = 0;
    const buffer: Uint8Array[] = [];

    function onAborted() {
      if (complete) return;
      done(httpError(400, 'request aborted', {
        code: 'ECONNABORTED',
        expected: length,
        length: length,
        received: received,
        type: 'request.aborted',
      }));
    }

    function onData(chunk: any) {
      if (complete) return;
      received += chunk.length;
      if (limit && received > limit) {
        done(httpError(413, 'request entity too large', {
          limit: limit,
          received: received,
          type: 'entity.too.large',
        }));
      } else {
        buffer.push(chunk);
      }
    }
  
    function onEnd(err: any) {
      if (complete) return;
      if (err) return done(err);
      if (length && received !== length) {
        done(httpError(400, 'request size did not match content length', {
          expected: length,
          length: length,
          received: received,
          type: 'request.size.invalid',
        }));
      } else {
        done(Buffer.concat(buffer));
      }
    }

    // attach listeners
    stream.on('aborted', onAborted);
    stream.on('close', cleanup);
    stream.on('data', onData);
    stream.on('end', onEnd);
    stream.on('error', onEnd);

    // 清理
    function cleanup() {
      stream.removeListener('aborted', onAborted);
      stream.removeListener('data', onData);
      stream.removeListener('end', onEnd);
      stream.removeListener('error', onEnd);
      stream.removeListener('close', cleanup);
      stream.unpipe();
    }

    // 完成后调用
    function done(arg: Error | Buffer) {
      complete = true;
      cleanup();
      if (arg instanceof Error) {
        reject(arg);
      } else {
        resolve(arg);
      }
    }
  });
}


src/index.ts
import { SetupFunction } from '@zenweb/core';
import { JSONParser, RawBodyParser, TextBodyParser, URLEncodedParser } from './parse';
import { BodyOption } from './types';
export * from './types';
export * from './body';
export * from './parse';

const defaultOption: BodyOption = {
  encoding: 'utf-8',
  limit: 1024 * 1024,
  inflate: true,
  textTypes: ['text/*', 'xml', '+xml'],
  parses: [JSONParser, URLEncodedParser],
};

export default function setup(opt?: BodyOption): SetupFunction {
  opt = Object.assign({}, defaultOption, opt);
  return function body(setup) {
    setup.debug('option: %o', opt);
    setup.assertModuleExists('inject');
    setup.assertModuleExists('helper');
    setup.core.injector.define(BodyOption, opt);
    // 在初始化后期执行解析器载入工作，方便其他模块添加解析器
    setup.after(async () => {
      // 载入解析器
      if (opt?.parses) {
        if (!opt.textTypes) opt.textTypes = [];
        for (const parserClass of opt.parses) {
          setup.debug('init parser: %o', parserClass);
          const parser = await setup.core.injector.getInstance(parserClass);
          if (parser instanceof TextBodyParser) {
            setup.debug('add textTypes: %o', parser.types);
            for (const type of parser.types) {
              if (!opt.textTypes.includes(type)) {
                opt.textTypes.push(type);
              }
            }
          } else if (parser instanceof RawBodyParser) {
          } else {
            throw new Error('unknonw parser type: ' + parser);
          }
        }
        setup.debug('all textTypes: %o', opt.textTypes);
      }
    });
  }
}


src/parse.ts
import { scope } from '@zenweb/inject';
import * as httpError from 'http-errors';
import * as querystring from 'querystring';

/**
 * 请求数据解析器基类
 */
@scope('singleton')
export abstract class BodyParser {
  /**
   * 是否为对象化的数据，可以被转换为 `ObjectBody`
   */
  abstract objected: boolean;

  /**
   * 支持的 mimetype 类型，使用 `ctx.is` 判断
   */
  abstract types: string[];

  /**
   * 解析处理方法
   */
  abstract parse(data: Buffer | string): unknown | Promise<unknown>;

  /**
   * 处理失败调用异常
   */
  fail(...args: httpError.UnknownError[]): never {
    throw httpError(...args);
  }
}

/**
 * 原始数据解析
 */
export abstract class RawBodyParser extends BodyParser {
  /**
   * 解析处理方法
   * @param data 原始数据
   */
  abstract parse(data: Buffer): unknown | Promise<unknown>;
}

/**
 * 文本数据解析 - 经过编码转换后的文本数据
 */
export abstract class TextBodyParser extends BodyParser {
  /**
   * 解析处理方法
   * @param text 编码转换后的文本数据
   */
  abstract parse(text: string): unknown | Promise<unknown>;
}

// Allowed whitespace is defined in RFC 7159
// http://www.rfc-editor.org/rfc/rfc7159.txt
const strictJSONReg = /^[\x20\x09\x0a\x0d]*(\[|\{)/;

/**
 * JSON 解析处理
 */
export class JSONParser extends TextBodyParser {
  objected = true;
  types = ['json', '+json'];

  parse(text: string) {
    if (!strictJSONReg.test(text)) {
      this.fail(415, 'invalid JSON, only supports object or array', {
        type: 'json.strict',
      });
    }
    try {
      return JSON.parse(text);
    } catch (err: any) {
      this.fail(400, err.message, {
        type: 'json.parse-error',
      });
    }
  }
}

/**
 * form-urlencoded 表单解析
 */
export class URLEncodedParser extends TextBodyParser {
  objected = true;
  types = ['urlencoded'];

  parse(text: string) {
    return querystring.parse(text);
  }
}


src/body.ts
import { Context } from '@zenweb/core';
import { init, inject, scope } from '@zenweb/inject';
import { TypeCastHelper } from '@zenweb/helper';
import * as iconv from 'iconv-lite';
import * as httpError from 'http-errors';
import { BodyOption } from './types';
import { TypeCastPickOption } from 'typecasts';
import { streamReader } from './read';
import { BodyParser, RawBodyParser, TextBodyParser } from './parse';

/**
 * 原始请求内容-经过解压，未经过文字编码转换
 */
@scope('request')
export class RawBody {
  data?: Buffer;

  @init
  private async [Symbol()](option: BodyOption, ctx: Context) {
    if (!ctx.request.length) {
      return;
    }
    this.data = await streamReader(ctx, option.limit, option.inflate);
  }
}

/**
 * 文本内容，经过文字编码转换
 */
@scope('request')
export class TextBody {
  data?: string;

  // 再解析数据
  @init
  private async [Symbol()](option: BodyOption, ctx: Context) {
    // 是否是支持的类型
    if (!option.textTypes?.length || !ctx.is(option.textTypes)) {
      // throw httpError(415, 'unsupported type "' + ctx.request.type + '"', {
      //   type: 'type.unsupported',
      // });
      return;
    }
    // 是否满足编码转换
    const encoding = ctx.request.charset || option.encoding || 'utf-8';
    if (!iconv.encodingExists(encoding)) {
      throw httpError(415, 'unsupported charset "' + encoding.toUpperCase() + '"', {
        charset: encoding.toLowerCase(),
        type: 'charset.unsupported',
      });
    }
    // 读取数据
    const raw = await ctx.injector.getInstance(RawBody);
    if (!raw.data) {
      return;
    }
    // 编码转换
    try {
      this.data = iconv.decode(raw.data, encoding);
    } catch {
      throw httpError(400, 'decode charset failed', {
        charset: encoding.toLowerCase(),
        type: 'decode.failed',
      });
    }
  }
}

/**
 * 请求 Body 数据解析 - 混合解析器
 */
@scope('request')
export class Body {
  /**
   * 解析出的数据
   */
  data?: unknown;

  /**
   * 数据类型
   * - 如果匹配解析器中支持的类型则返回匹配的类型
   * - 如果没有匹配的解析器，但是匹配了 `textTypes` 则统一为 `text` 类型
   */
  type?: 'text' | string;

  /**
   * 匹配的解析器
   */
  parser?: BodyParser;

  @init
  private async [Symbol()](opt: BodyOption, ctx: Context) {
    // 匹配内容解析器
    if (opt.parses) {
      for (const parserClass of opt.parses) {
        const parser = await ctx.injector.getInstance(parserClass);
        const type = ctx.is(parser.types);
        if (type) {
          this.type = type;
          this.parser = parser;
          if (parser instanceof TextBodyParser) {
            const textBody = await ctx.injector.getInstance(TextBody);
            if (textBody.data) {
              this.data = await parser.parse(textBody.data);
            }
          } else if (parser instanceof RawBodyParser) {
            const rawBody = await ctx.injector.getInstance(RawBody);
            if (rawBody.data) {
              this.data = await parser.parse(rawBody.data);
            }
          }
          return;
        }
      }
    }

    // 没有匹配的解析器则尝试解析文本类型
    if (opt.textTypes && ctx.is(opt.textTypes)) {
      this.data = (await ctx.injector.getInstance(TextBody)).data;
      this.type = 'text';
      return;
    }

    // 2023-3-19: 
    // - 不解析为 raw。对于文件上传和Body共同使用的场景下，一般来说是不需要再读取流的，浪费内存，如果真的需要直接使用 RawBody 即可
    // const raw = await ctx.injector.getInstance(RawBody);
    // if (raw.data) {
    //   this.data = raw.data;
    //   this.type = 'raw';
    // }
  }
}

/**
 * 请求 Body 数据解析为对象本身，请求内容必须为 json 或 form-urlencoded
 * - 注意：数据不是必须的，但是如果传递的数据是不能被正确解析为对象化的将会抛出异常
 */
@scope('request')
export class ObjectBody {
  [key: string]: unknown;

  @init
  private async [Symbol()](body: Body) {
    if (!body.data) return;
    if (!body.parser?.objected) {
      throw httpError(400, 'only supports objected format: JSON or urlencoded', {
        type: 'objected.only',
      });
    }
    Object.assign(this, body.data);
  }
}

/**
 * ObjectBody 数据类型转换与校验
 */
@scope('request')
export class BodyHelper {
  @inject typeCastHelper!: TypeCastHelper;
  @inject data!: ObjectBody;

  get<O extends TypeCastPickOption>(fields: O) {
    return this.typeCastHelper.pick(this.data, fields);
  }
}
